cloudbot/util/decorators.py

```diff
@@ -0,0 +1,25 @@
+from time import time
+
+SPAM_LIST = {}
+
+
+def spam_guard(func):
+    def wrapper(*args, **kwargs):
+        nick = kwargs.get('nick', None)  # pull the nick from the args
+        if nick:  # if there is no nick specified for whatever reason, carry on
+            global SPAM_LIST
+            if SPAM_LIST.get(nick, None):
+                SPAM_LIST[nick][0] = time()
+                if time() - SPAM_LIST[nick][0] < SPAM_LIST[nick][1]:
+                    # notice that they are spamming
+                    SPAM_LIST[nick][1] += 10
+                    return (
+                        "{} attempted to execute a command within their previous" \
+                        " limit. An additional 10 seconds has been added. Current" \
+                        " limit is {} seconds.".format(nick, SPAM_LIST[nick][1])
```

Comment:
1. After a bit of tweaking with the decorator to make it work, I was able to get the following output:  
  
```
Feb 08 16:33:25 <FxChiP>        sessions
Feb 08 16:33:27 <FxChiP>        joinSession 1
Feb 08 16:33:27 <FxLight>       (FxChiP) FxChiP attempted to execute a command within their previous limit. An additional 10 seconds has been added. Current limit is 20 seconds.
```

This is in PM, but the results would be the same for a channel. 

2. Turns out the bot has a built-in rate-limiting plugin for commands, see https://github.com/CloudBotIRC/CloudBot/wiki/Configuration . You assign a token cost for messages, an initial token value for users, and a refill rate of tokens over time; when the user uses a command, they expend the message token cost and must wait for the refill to use another command. I really wish CloudBot mainline would have documented this better, but it *is* here.